File: AtlasWeb/Core/Modules/Tibbi_Surec/Patoloji_Modulu/PathologyMainFormViewModel.ts
//$2F31D9F8
import { BaseViewModel } from "NebulaClient/Model/BaseViewModel";
import { Pathology } from 'NebulaClient/Model/AtlasClientModel';
import { DiagnosisGrid } from 'NebulaClient/Model/AtlasClientModel';
import { PathologyMaterial } from 'NebulaClient/Model/AtlasClientModel';
import { BaseTreatmentMaterial } from 'NebulaClient/Model/AtlasClientModel';
import { PathologyRequest } from 'NebulaClient/Model/AtlasClientModel';
import { ResUser } from 'NebulaClient/Model/AtlasClientModel';
import { Episode } from 'NebulaClient/Model/AtlasClientModel';
import { DiagnosisDefinition } from 'NebulaClient/Model/AtlasClientModel';
import { SKRSICDOYERLESIMYERI } from 'NebulaClient/Model/AtlasClientModel';
import { SKRSNumuneAlindigiDokununTemelOzelligi } from 'NebulaClient/Model/AtlasClientModel';
import { SKRSNumuneAlinmaSekli } from 'NebulaClient/Model/AtlasClientModel';
import { SKRSPatolojiPreparatiDurumu } from 'NebulaClient/Model/AtlasClientModel';
import { SKRSICDOMORFOLOJIKODU } from 'NebulaClient/Model/AtlasClientModel';
import { Material } from 'NebulaClient/Model/AtlasClientModel';
import { StockCard } from 'NebulaClient/Model/AtlasClientModel';
import { DistributionTypeDefinition } from 'NebulaClient/Model/AtlasClientModel';
import { PathologyTestProcedure } from 'NebulaClient/Model/AtlasClientModel';
import { Guid } from 'NebulaClient/Mscorlib/Guid';
import { Type } from 'NebulaClient/ClassTransformer';
import { vmRequestedPathologyProcedure } from './PathologyMaterialInfoViewModel';
import { PathologyAdditionalReport } from 'NebulaClient/Model/AtlasClientModel';
import { PathologyPanicAlert } from "NebulaClient/Model/AtlasClientModel";

export class PathologyMainFormViewModel extends BaseViewModel {
    protected __type__: string = 'Core.Models.PathologyMainFormViewModel, Core';

    @Type(() => Pathology)
    public _Pathology: Pathology = new Pathology();
    @Type(() => PathologyAdditionalReport)
    public PathologyAdditionalReportGridList: Array<PathologyAdditionalReport> = new Array<PathologyAdditionalReport>();
    @Type(() => PathologyPanicAlert)
    public PathologyPanicAlertGridList: Array<PathologyPanicAlert> = new Array<PathologyPanicAlert>();
    @Type(() => DiagnosisGrid)
    public DiagnosisDiagnosisGridGridList: Array<DiagnosisGrid> = new Array<DiagnosisGrid>();
    @Type(() => PathologyMaterial)
    public PathologyMaterialGridList: Array<PathologyMaterial> = new Array<PathologyMaterial>();
    @Type(() => BaseTreatmentMaterial)
    public TreatmentMaterialsGridList: Array<BaseTreatmentMaterial> = new Array<BaseTreatmentMaterial>();
    @Type(() => PathologyRequest)
    public PathologyRequests: Array<PathologyRequest> = new Array<PathologyRequest>();
    @Type(() => ResUser)
    public ResUsers: Array<ResUser> = new Array<ResUser>();
    @Type(() => Episode)
    public Episodes: Array<Episode> = new Array<Episode>();
    @Type(() => DiagnosisDefinition)
    public DiagnosisDefinitions: Array<DiagnosisDefinition> = new Array<DiagnosisDefinition>();
    @Type(() => SKRSICDOYERLESIMYERI)
    public SKRSICDOYERLESIMYERIs: Array<SKRSICDOYERLESIMYERI> = new Array<SKRSICDOYERLESIMYERI>();
    @Type(() => SKRSNumuneAlindigiDokununTemelOzelligi)
    public SKRSNumuneAlindigiDokununTemelOzelligis: Array<SKRSNumuneAlindigiDokununTemelOzelligi> = new Array<SKRSNumuneAlindigiDokununTemelOzelligi>();
    @Type(() => SKRSNumuneAlinmaSekli)
    public SKRSNumuneAlinmaSeklis: Array<SKRSNumuneAlinmaSekli> = new Array<SKRSNumuneAlinmaSekli>();
    @Type(() => SKRSPatolojiPreparatiDurumu)
    public SKRSPatolojiPreparatiDurumus: Array<SKRSPatolojiPreparatiDurumu> = new Array<SKRSPatolojiPreparatiDurumu>();
    @Type(() => SKRSICDOMORFOLOJIKODU)
    public SKRSICDOMORFOLOJIKODUs: Array<SKRSICDOMORFOLOJIKODU> = new Array<SKRSICDOMORFOLOJIKODU>();
    @Type(() => Material)
    public Materials: Array<Material> = new Array<Material>();
    @Type(() => StockCard)
    public StockCards: Array<StockCard> = new Array<StockCard>();
    @Type(() => DistributionTypeDefinition)
    public DistributionTypeDefinitions: Array<DistributionTypeDefinition> = new Array<DistributionTypeDefinition>();
    @Type(() => MaterialProcedure)
    public MaterialProcedureList: Array<MaterialProcedure> = new Array<MaterialProcedure>();
    @Type(() => vmRequestedPathologyProcedure)
    public _RequestedPathologyProceduresArray: Array<vmRequestedPathologyProcedure> = new Array<vmRequestedPathologyProcedure>();
    public _StateName: string = "";
    public _PatientObjectID: string;
    public hasRolePathologist: boolean;
}

export class MaterialProcedure {
    @Type(() => Guid)
    public MaterialObjectID: Guid;
    @Type(() => PathologyTestProcedure)
    public PathologyTestProcedureGridList: Array<PathologyTestProcedure> = new Array<PathologyTestProcedure>();
}

export class PathologySubepisodeInfo {
    public SubepisodeID: string;
    public ProtocolNo: string;
    public SubepisodeOpeningDate: string;
    public RessectionName: string;
    public DoctorName: string;
    @Type(() => PathologyInfo)
    public pathologyInfos: Array<PathologyInfo> = new Array<PathologyInfo>();
}

export class PathologyInfo {
    public PathologyProtocolNo: string;
    public ApproveDate: string;
    public ReportDate: string;
    public PathologyID: string;
    public CurrentStateName: string;
    public CurrentStateID: string;
    public PatDoctorName: string;
    public IsApproved:boolean;
    public IsReported:boolean;
}

File: AtlasWeb/Core/Modules/Tibbi_Surec/Patoloji_Modulu/PathologyMainFormViewModel.cs
//$06277CC6
using System;
using System.Linq;
using TTInstanceManagement;
using Core.Models;
using TTObjectClasses;
using TTDefinitionManagement;
using TTUtils;
using System.Collections.Generic;
using Infrastructure.Models;
using Microsoft.AspNetCore.Mvc;

namespace Core.Controllers
{
    public partial class PathologyServiceController
    {
        partial void PostScript_PathologyMainForm(PathologyMainFormViewModel viewModel, Pathology pathology, TTDefinitionManagement.TTObjectStateTransitionDef transDef, TTObjectContext objectContext)
        {
            //TANI için
            pathology.DiagnosisGrid_PostScript(viewModel.DiagnosisDiagnosisGridGridList);

            foreach (vmRequestedPathologyProcedure procedure in viewModel._RequestedPathologyProceduresArray)
            {
                var importedMaterial = (PathologyMaterial)objectContext.GetLoadedObject(procedure.MaterialObjectID);
                if (importedMaterial.CurrentStateDefID != PathologyMaterial.States.Reject)
                {
                    if (procedure.SubActionProcedureObjectId == Guid.Empty)
                    {
                        var pathologyTest = new PathologyTestProcedure(objectContext);
                        var pathologyDef = (PathologyTestDefinition)objectContext.GetObject(procedure.ProcedureDefObjectID, "PathologyTestDefinition");
                        pathologyTest.ProcedureObject = pathologyDef;
                        pathologyTest.Amount = procedure.Amount;
                        pathologyTest.EpisodeAction = pathology;
                        pathologyTest.PathologyMaterial = importedMaterial;
                        pathologyTest.PathologyRequest = pathology.PathologyRequest;
                        pathologyTest.ProcedureDoctor = pathology.ProcedureDoctor;
                        pathologyTest.ActionDate = DateTime.Now;
                        pathologyTest.CreationDate = procedure.RequestDate;
                        pathologyTest.CurrentStateDefID = PathologyTestProcedure.States.New;
                    }
                }
             
            }
            objectContext.Save();
        }

        partial void PreScript_PathologyMainForm(PathologyMainFormViewModel viewModel, Pathology pathology, TTObjectContext objectContext)
        {
            viewModel._StateName = pathology.CurrentStateDef.DisplayText;
            viewModel._PatientObjectID = pathology.Episode.Patient.ObjectID.ToString();
            // SARF GRIDINDE Cancelledlar gelmemesi  için  Bu koddan sonra ContextToViewModel çalýþmamalý
            viewModel.TreatmentMaterialsGridList = viewModel.TreatmentMaterialsGridList.Where(dr => dr.CurrentStateDef.Status != StateStatusEnum.Cancelled).ToArray();

            //ContextToViewModel den sonra çaðýrýlmalý //TANI için
            viewModel.DiagnosisDiagnosisGridGridList = pathology.DiagnosisGrid_PreScript();

            if (Common.CurrentUser.HasRole(TTRoleNames.Patoloji_Uzmani) == true)
            {
                viewModel.hasRolePathologist = true;
            }

            else
            {
                viewModel.hasRolePathologist = false;
            }

            if (viewModel._Pathology.CurrentStateDefID == Pathology.States.Procedure && Common.CurrentResource.TakesPerformanceScore == true && viewModel._Pathology.ProcedureDoctor == null)
            {
                viewModel._Pathology.ProcedureDoctor = Common.CurrentResource;
                
            }

        }

        [HttpGet]
        public string CheckAdditionalReport(Guid PathologyID)
        {
            using (var objectContext = new TTObjectContext(false))
            {

                Pathology pathology = objectContext.GetObject<Pathology>(PathologyID);
                if (pathology.PathologyAdditionalReport != null)
                {
                    return pathology.PathologyAdditionalReport.ObjectID.ToString();
                }
                return "";

            }

        }

        [HttpGet]
        public string CheckUnapprovedPathologyReports(Guid PathologyID)
        {
            string result = "";
            using (var objectContext = new TTObjectContext(true))
            {
                Pathology pathology = objectContext.GetObject<Pathology>(PathologyID);

                List<Pathology.GetUnapprovedPathologyReports_Class> reports = Pathology.GetUnapprovedPathologyReports(pathology.Episode.Patient.ObjectID,pathology.ObjectID).ToList();
                if(reports.Count>0)
                {
                   
                    foreach (Pathology.GetUnapprovedPathologyReports_Class report in reports)
                    {

                        Pathology pathologyReport = objectContext.GetObject<Pathology>(new Guid(report.Pathologyid.ToString()));

                        int resultTime = 0;
                        foreach (PathologyMaterial material in pathologyReport.PathologyMaterial)
                        {
                            
                            foreach (PathologyTestProcedure test in material.PathologyTestProcedure) //Rapordaki patoloji iþlemlerinin arasýndaki en uzun sürede sonuçlanan iþlemin sonuçlanma zamaný
                            {
                                if(((PathologyTestDefinition)test.ProcedureObject).ResultTime != null)
                                {
                                    if (Convert.ToInt32(((PathologyTestDefinition)test.ProcedureObject).ResultTime) > resultTime)
                                        resultTime = Convert.ToInt32(((PathologyTestDefinition)test.ProcedureObject).ResultTime); 
                                    

                                }
                                
                            }
                        }

                        if(resultTime != 0 && result == "" && Convert.ToDateTime(pathologyReport.PathologyRequest.AcceptionDate).AddDays(resultTime) < DateTime.Now)
                        {
                            result = "Hastaya Ait Onaylanmamýþ Patoloji Raporlarý Bulunmaktadýr. Protokol Numaralarý; ";
                            result += "<br/>" + report.MatPrtNoString;
                        }else if(resultTime != 0 && result != ""  && Convert.ToDateTime(pathologyReport.PathologyRequest.AcceptionDate).AddDays(resultTime) < DateTime.Now)
                        {
                            result += "<br/>" + report.MatPrtNoString;
                        }


                    }
                }
           

            }
            return result;
        }


        [HttpGet]
        public string CheckPanicAlert(Guid PathologyID)
        {
            using (var objectContext = new TTObjectContext(false))
            {

                Pathology pathology = objectContext.GetObject<Pathology>(PathologyID);
                if (pathology.PathologyPanicAlert !=  null)
                {
                    return pathology.PathologyPanicAlert.ObjectID.ToString();
                }
                return "";

            }

        }
    }
}

namespace Core.Models
{
    public partial class PathologyMainFormViewModel : BaseViewModel
    {
        public MaterialProcedure[] MaterialProcedureList
        {
            get;
            set;
        }

        public vmRequestedPathologyProcedure[] _RequestedPathologyProceduresArray
        {
            get;
            set;
        }

        public string _StateName
        {
            get;
            set;
        }

        public string _PatientObjectID
        {
            get;
            set;
        }

        public bool hasRolePathologist
        {
            get;
            set;
        }
    }

    public class MaterialProcedure
    {
        public Guid MaterialObjectID;
        public TTObjectClasses.PathologyTestProcedure[] PathologyTestProcedureGridList
        {
            get;
            set;
        }
    }
}